internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go

```diff
@@ -30,14 +34,48 @@ func NewUpgrade(log *logger.Logger, coord *coordinator.Coordinator) *Upgrade {
 	}
 }
 
-// Handle handles UPGRADE action.
-func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, _ acker.Acker) error {
+// Handle handles UPGRADE action.  Returns immediately and the actual
+// upgrade happens asynchronously.  This allows for downloads to
+// happen without blocking updates.  If multiple upgrades are sent
+// then we ack them all if there is an error, but only the first actually executes.
+// If successful, reboot does ACK and check-in.
+func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker) error {
 	h.log.Debugf("handlerUpgrade: action '%+v' received", a)
 	action, ok := a.(*fleetapi.ActionUpgrade)
 	if !ok {
 		return fmt.Errorf("invalid type, expected ActionUpgrade and received %T", a)
 	}
+	go func() {
+		h.log.Infof("starting upgrade to version %s in background", action.Version)
+		h.addAction(a)
```

Comment:
I feel like the logic of ensuring that the ack should be queued should be in this function with if returning trye if it should even call `h.coord.Upgrade`.

I think this should include the following logic, as we discussed in the zoom call:

1. Check if the `actions` is empty, if its empty add to the array and start upgrade.
2. If `actions` is not empty ensure that this upgrade ack matches the same values in the array. If it does not match, it should cancel the running upgrade context, trigger reporting all the other ack's as errors, start a new upgrade process, adding this ack to the `actions` queue.

I feel like this logic is missing here, if I am incorrect please let me know.

Reply:
updated with cancel running upgrade, and start new upgrade if version numbers differ